lib/index.js

```diff
@@ -295,8 +310,156 @@ Segment.prototype.referrerId = function(query, ctx) {
   this.cookie('s:context.referrer', json.stringify(ad));
 };
 
+
 /**
- * Noop.
+ * retrieveCrossDomainId.
+ *
+ * @api private
+ * @param {function) callback => err, {crossDomainId, fromServer, timestamp}
+ */
+Segment.prototype.retrieveCrossDomainId = function(callback) {
+  if (!this.cookie('segment_cross_domain_id')) {
+    var self = this;
+    var writeKey = this.options.apiKey;
+    
+    // Exclude the current domain from the list of servers we're querying
+    var currentTld = getTld(window.location.hostname);
+    var domains = [];
+    for (var i=0; i<this.options.crossDomainIdServers.length; i++) {
+      var domain = this.options.crossDomainIdServers[i];
+      if (getTld(domain) !== currentTld) {
+        domains.push(domain);
+      }
+    }
+    
+    getCrossDomainIdFromServerList(domains, writeKey, function(err, res) {
+      if (err) {
+        // We optimize for no conflicting xid as much as possible. So bail out if there is an
+        // error and we cannot be sure that xid does not exist on any other domains
+        if (callback) {
+          callback(err, null);
+        }
+        return;
+      }
+      var crossDomainId = null;
+      var fromDomain = null;
+      if (res) {
+        crossDomainId = res.id;
+        fromDomain = res.domain;
+      } else {
+        crossDomainId = uuid();
+        fromDomain = window.location.hostname;
+      }
+      var currentTimeMillis = (new Date()).getTime();
+      self.cookie('segment_cross_domain_id', crossDomainId);
+      // Not actively used. Saving for future conflict resolution purposes
+      self.cookie('segment_cross_domain_id_from_domain', fromDomain);
```

Comment:
I know I recommended storing this as a cookie - but i wonder if we're not using it this would be better in local storage? To void running into size limits.